keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,578 @@
+  WorkloadAnnotation = "kueue.x-k8s.io/workload"
+
+  // PodSetLabel indicates the name of the PodSet in the workload
+  PodSeLabel = "kueue.x-k8s.io/podset"
+)
+```
+
+### Computing the assignment
+
+The extended pod assignment is set on admitting the workload. In order to
+compute the assignment Kueue caches the information about the node allocatable
+capacity and the current usage by other workloads in this resource flavor.
+
+The information about the allocatable capacity is scraped from nodes based on
+the `status.allocatable` field.
+
+The cached information about the used resources is updated whenever a workload
+is admitted, suspended, resumed or finished. Additionally, we scrape the usage
+information from all non-TAS pods bound to the TAS nodes - this is done to
+account for Pods not created by Workloads not managed by Kueue (such as
+DaemonSet pods).
+
+The available capacity for TAS nodes is then computed as a difference between
+the allocatable space and current usage.
+
+For a given PodSet Kueue:
+- when the `require-topology` is used, then Kueue tries to find any value of the
+ level label which can accommodate all the pods. If there is no such value, then
+ the workload keeps waiting in the queue.
+- when the `prefer-topology` is used, then Kueue tries to use heuristics to
+  minimize the number of values of the label used. One approach is to greedily
+  choose the value which can accommodate the most pods.
+
+### Enforcing the assignment
+
+When the workload has the PodSet assignments which are non-homogenous (all Pods
+landing in the same topology) and is about to start we modify the corresponding
+PodTemplates in the Job object to inject the `kueue.x-k8s.io/topology`
+scheduling gate. For homogenous PodSet assignments it is enough to inject the
+corresponding nodeSelector in the PodTemplates.
+
+Then, there is a new component, called `TopologyUngater`, which is a Workload
+reconciler which lists all pods for a given TAS PodSet, and ensures that the
+pods in the expected number are un-gated to a given value.
+
+Along with the scheduling gate, to each pod template the
+`kueue.x-k8s.io/workload` and `kueue.x-k8s.io/podset` labels / annotations are
+added to facilitate quick lookup (List request) for all pods corresponding to
+the workload (and more specifically PodSetAssignment) by `TopologyAssigner`.
+
+The TopologyUngater watches for pod events which trigger the reconciliation
+loop. The reconciliations are batched by 1s periods to make sure multiple
+created pods at the similar time don't trigger too many reconciliations.
+
+In order to prevent ungating more pods as expected for a PodSet we consider to
+use the expectations mechanism. The expectations are set for when we are about
+to ungate a Pod. The expectation is fulfield if the Pod is observed as ungated
+or the ungating request fails. We hold ungating if there are pending ungatings
+within the PodSet.
+
+### Support for ReplicatedJobs in JobSet
```

Comment:
There is still no explanation of where the user defines the annotations and what do they mean for jobset.

Reply:
Good point, I was a bit hesitant at which level we should add the labels. Added examples to demonstrate the current design. I admit there is no option in the proposal to control the two:
1. ability to specify a different topology domain per PodTemplate
2. ability to specify a topology domain for the entire collection of Jobs (one may want to say "require: block")

I think these are still possible with API extensions if we need them in the future. 

For (1.) we could have labels also at the PodTemplate level (then we would probably need to say - either specify at the PodTemplate level.
For (2,) we could introduce a new label like "topology-require-for-all-pods: block". We probably don't need "prefer" at the top level, as it would be implied by "prefer" at the Jobs level.

Reply:
wdyt @mwielgus ?

Reply:
Marking as resolved since we decided not to support the ReplicatedJob level for Alpha. Moved the started effort to alternatives.